Sources/tasks/ExportProfileData.php

```diff
@@ -0,0 +1,253 @@
+		$user_info = $member_info;
+
+		loadEssentialThemeData();
+		$settings['actual_theme_dir'] = $settings['theme_dir'];
+		$context['user']['language'] = $lang;
+		loadMemberData($uid);
+		loadLanguage(implode('+', array_unique(array('index', 'Modifications', 'Stats', 'Profile', $included[$datatype]['langfile']))), $lang);
+
+		// @todo Ask lawyers whether the GDPR requires us to include posts in the recycle bin.
+		$query_this_board = '{query_see_board}' . (!empty($modSettings['recycle_enable']) && $modSettings['recycle_board'] > 0 ? ' AND b.id_board != ' . $modSettings['recycle_board'] : '');
+
+		// We need a valid export directory.
+		if (empty($modSettings['export_dir']) || !file_exists($modSettings['export_dir']))
+		{
+			require_once($sourcedir . DIRECTORY_SEPARATOR . 'Profile-Actions.php');
+			if (create_export_dir() === false)
+				return;
+		}
+
+		$export_dir_slash = $modSettings['export_dir'] . DIRECTORY_SEPARATOR;
+
+		$idhash = hash_hmac('sha1', $uid, get_auth_secret());
+		$idhash_ext = $idhash . '.xml';
+
+		// Increment the file number until we reach one that doesn't exist.
+		$filenum = 1;
+		$realfile = $export_dir_slash . $filenum . '_' . $idhash_ext;
+		while (file_exists($realfile))
+			$realfile = $export_dir_slash . ++$filenum . '_' . $idhash_ext;
+
+		$tempfile = $export_dir_slash . $idhash_ext . '.tmp';
+		$progressfile = $export_dir_slash . $idhash_ext . '.progress.json';
+
+		$feed_meta = array(
+			'title' => sprintf($txt['profile_of_username'], $user_profile[$uid]['real_name']),
+			'desc' => sentence_list(array_map(function ($datatype) use ($txt) { return $txt[$datatype]; }, array_keys($included))),
+			'author' => $mbname,
+			'source' => $scripturl . '?action=profile;u=' . $uid,
+			'self' => $scripturl . '?action=profile;area=download;u=' . $uid . ';t=' . hash_hmac('sha1', $idhash, get_auth_secret()),
+		);
+
+		// If a necessary file is missing, we need to start over.
+		if (!file_exists($progressfile) || !file_exists($tempfile))
+		{
+			foreach (array_merge(array($tempfile, $progressfile), glob($export_dir_slash . '*_' . $idhash_ext)) as $fpath)
+				@unlink($fpath);
+
+			buildXmlFeed('smf', array(), $feed_meta, 'profile');
+			file_put_contents($tempfile, implode('', array($context['feed']['header'], $context['feed']['footer'])));
+		}
+
+		$progress = file_exists($progressfile) ? $smcFunc['json_decode'](file_get_contents($progressfile), true) : array_fill_keys($datatypes, 0);
+
+		// Get the data, always in ascending order.
+		$xml_data = call_user_func($included[$datatype]['func'], 'smf', true);
+
+		// Build the XML string from the data.
+		buildXmlFeed('smf', $xml_data, $feed_meta, $datatype);
+
+		$last_item = end($xml_data);
+		if (isset($last_item['content'][0]['content']) && $last_item['content'][0]['tag'] === 'id')
+			$last_id = $last_item['content'][0]['content'];
+
+		// Append the string (assuming there's enough disk space).
+		$diskspace = disk_free_space($modSettings['export_dir']);
```

Comment:
And a note: It's _there_, it just doesn't work.  It may return null, or 0, or 1...

Reply:
Hm. It never occurred to me that the function could be enabled but rigged to return a fudged value.

But making sure exports don't consume all available disk space is a much less theoretical concern than making sure there is enough disk space to write an updated Settings.php file (as in #6061), so I don't think we should just forget about the need for this. I'll have to ponder it a bit more...

Reply:
Problem is neither are theoretical.  That null returned value in #6061 was detected only a couple days after that PR went in.  We can't assume it's usable.  

I don't think we have an acceptable way to detect available space in advance that will work on all hosts.  

However, a write error will be caught by error trapping after the write attempt.  

If your disks are full, this is going to be the least of your worries...  We could possibly halt export attempts if we see a disk full returned.

Reply:
> Problem is neither are theoretical. [...]

I think you misunderstood me. I agree with you that it creates a real problem if the disk space functions are missing. Rather, when I referred to more and less theoretical issues, I was contrasting the severity of the implications of that problem in the two different cases. In the case of Settings.php it is fine to deal with the problem of the missing functions by simply skipping the `disk_free_space()` check, since changing settings presents no risk of consuming vast quantities of disk space. In contrast, exports can get big quickly and can be initiated by users, which means it is very important to give the admin the ability to monitor and control disk usage in relation to exports. But not having access to information about free space on disk makes that kind of monitoring and control virtually impossible, so the consequences of those functions being disabled are much more severe in the case of exports than in the case of Settings.php. That's the contrast that I am making.